File: Sorting Techniques/Merge Sort/MergeSort.java
import java.util.Arrays;

public class MergeSort {
    public int[] mergeSort(int[] arr, int st, int ed){
         if (st == ed) {
            return new int[] {arr[st]};  // Base case: an array of size 1 is already sorted
        }
        
        int mid = st + (ed - st)/2;

        int[] left = mergeSort(arr, st, mid);
        int[] right = mergeSort(arr, mid+1, ed);

        int[] res = mergeArray(left, right);

        return res;
    }

    public int[] mergeArray(int[] arr1, int[] arr2){
        int[] res = new int[arr1.length + arr2.length];
        
        int i = 0;
        int j = 0;
        int k = 0;

        while(i < arr1.length && j < arr2.length){
            if(arr1[i] <= arr2[j]){
                res[k] = arr1[i];
                i++;
                k++;
            } else{
                res[k] = arr2[j];
                j++;
                k++;
            }
        }

        while(i < arr1.length){
            res[k] = arr1[i];
            i++;
            k++;
        }

        while(j < arr2.length){
            res[k] = arr2[j];
            j++;
            k++;
        }

        return res;
    }

    public static void main(String[] args){
        MergeSort sol = new MergeSort();

        int[] arr = {5,1,4,4,1};
        int st = 0;
        int ed = 4;

        int[] sortedArr = sol.mergeSort(arr, st, ed);

        System.out.println(Arrays.toString(sortedArr));
    }
}

File: Sorting Techniques/Merge Sort/MergeSort.cpp
#include<iostream>
#include <vector>

using namespace std;

class Solution{
public:
    vector<int> mergeSort(vector<int>& arr, int st, int ed){
        // Base case: an array of size 1 is already sorted
        if(st == ed){
            return {arr[st]};
        }

        int mid = st + (ed-st)/2;

        vector<int> left = mergeSort(arr, st, mid); //recursively solve the left part
        vector<int> right = mergeSort(arr, mid+1, ed); //recursively solve the right part

        return mergeArray(left, right); //merge the sorted left and right parts
    }

    vector<int> mergeArray(const vector<int>& arr1, const vector<int>& arr2){
        vector<int> res(arr1.size() + arr2.size());

        int i = 0;
        int j = 0;
        int k = 0;

        while(i < arr1.size() && j < arr2.size()){
            if(arr1[i] <= arr2[j]){
                res[k] = arr1[i];
                i++;
                k++;
            } else{
                res[k] = arr2[j];
                j++;
                k++;
            }
        }

        // Handle remaining elements of arr1
        while(i < arr1.size()){
            res[k] = arr1[i];
            i++;
            k++;
        }

        // Handle remaining elements of arr2
        while(j < arr2.size()){
            res[k] = arr2[j];
            j++;
            k++;
        }

        return res;
    }
};

int main(){
    Solution sol;

    vector<int> arr = {5, 1, 4, 4, 1};
    int st = 0;
    int ed = arr.size() - 1;  // Adjust the end index to the last element in the array

    vector<int> sortedArr = sol.mergeSort(arr, st, ed);

    // Output the sorted array
    for (int i : sortedArr) {
        cout << i << ", ";
    }

    return 0;
}